README.md
# datastructures
PHP extensions for classes supporting common datastructures

PHP 8.0 and higher

Current Structures:
1. Structs (Generic and Defined)
2. Enum

[![Build Status](https://travis-ci.org/eosforphp/datastructures.svg?branch=master)](https://travis-ci.org/eosforphp/datastructures)


CHANGELOG
2021-08-11  Swen Zanon  <swen.zanon@geoglis.de>
    * Ported to PHP8

2015-06-30  Elizabeth M Smith  <auroraeosrose@php.net>
    * Initial Extension
    * Enum Class


src/struct.c
/*
  +----------------------------------------------------------------------+
  | For PHP Version 8                                                    |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015-2017 Elizabeth M Smith                            |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Author: Elizabeth M Smith <auroraeosrose@gmail.com>                  |
  +----------------------------------------------------------------------+
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <php.h>
#include <zend_exceptions.h>

#include "php_eos_datastructures.h"

/**
 * Struct
 * Immutable\Struct
 * Typed\Struct
 */
zend_class_entry *ce_eos_datastructures_struct;
zend_class_entry *ce_eos_datastructures_immutable_struct;
zend_class_entry *ce_eos_datastructures_typed_struct;

static zend_object_handlers eos_datastructures_struct_object_handlers;
static zend_object_handlers eos_datastructures_immutable_struct_object_handlers;
static zend_object_handlers eos_datastructures_typed_struct_object_handlers;



static zval *eos_datastructure_struct_object_write_property(zend_object *object, zend_string *member, zval *value, void **cache_slot);

/* ----------------------------------------------------------------
    Eos\Datastructures\Struct C API
------------------------------------------------------------------*/

/* {{{ allows access to the struct class entry via other extensions */
PHP_EOS_DATASTRUCTURES_API zend_class_entry *php_eos_datastructures_get_struct_ce(void)
{
	return ce_eos_datastructures_struct;
}
/* }}} */

/* ----------------------------------------------------------------
    Eos\DataStructures\Struct class API
------------------------------------------------------------------*/

ZEND_BEGIN_ARG_INFO_EX(Struct___construct_args, ZEND_SEND_BY_VAL, ZEND_RETURN_VALUE, 0)
	ZEND_ARG_INFO(0, values)
ZEND_END_ARG_INFO()

/* {{{ proto void Struct->__construct(array/object values)
                  constructor for struct class, will walk hashtable key/values to set properties */
PHP_METHOD(EosDataStructuresStruct, __construct)
{
	HashTable *values = NULL;
	zend_string *str_idx;
	zval *entry;

        ZEND_PARSE_PARAMETERS_START(0,1)
                Z_PARAM_OPTIONAL
                Z_PARAM_ARRAY_HT(values)
        ZEND_PARSE_PARAMETERS_END();

	/**
	 * Our arguments hashtable has two different behaviors
	 * If we are in the generic struct class, we set each item
	 * as a public property
	 *
	 * If we are in an extended class, we set only existing properties
	 */
	if(values) {

		if(Z_OBJCE_P(getThis()) == ce_eos_datastructures_struct) {

			ZEND_HASH_FOREACH_STR_KEY_VAL(values, str_idx, entry) {
				if(str_idx) {
					(zend_get_std_object_handlers())->write_property(Z_OBJ_P(getThis()), str_idx, entry, NULL);
				}
			} ZEND_HASH_FOREACH_END();

		} else {

			/* mangle scope so we can get to private/protected */
			zend_class_entry *old_scope = EG(fake_scope);
			EG(fake_scope) = Z_OBJCE_P(getThis());

			ZEND_HASH_FOREACH_STR_KEY_VAL(values, str_idx, entry) {
				if(str_idx) {
					eos_datastructure_struct_object_write_property(Z_OBJ_P(getThis()), str_idx, entry, NULL);
				}
			} ZEND_HASH_FOREACH_END();

			/* return our scope to normal */
			EG(fake_scope) = old_scope;
		}
	}
}

/* ----------------------------------------------------------------
    Eos\DataStructures\Struct Object management
------------------------------------------------------------------*/

/* {{{ */
static zend_object* eos_datastructures_struct_create_object(zend_class_entry *ce)
{
	zend_object *object = ecalloc(1, sizeof(zend_object)
							 + zend_object_properties_size(ce));
	zend_object_std_init(object, ce);
	object_properties_init(object, ce);

	object->handlers = &eos_datastructures_struct_object_handlers;
	return object;
}
/* }}} */

/**
 * Helper for property exists lookup for limiting properties to declared
 * or declared with constructor for generics
 */
static inline zend_bool eos_datastructures_struct_property_exists(zend_object *object, zend_string *member) {
	int ret = 0;

	/**
	 * Generic structs are "property is already set"
	 */
	if(object->ce == ce_eos_datastructures_struct) {
		ret = (zend_get_std_object_handlers())->has_property(object, member, 0, NULL);
	} else {
		zend_property_info *property_info;
                
		/* mangle scope so we can get to private/protected */
		zend_class_entry *old_scope = EG(fake_scope);
		EG(fake_scope) = object->ce;

		property_info = zend_get_property_info(object->ce, member, 1);
		if(property_info!= ZEND_WRONG_PROPERTY_INFO &&
		   property_info &&
		   (property_info->flags & ZEND_ACC_STATIC) == 0) {
			ret = 1;
		}
		/* return our scope to normal */
		EG(fake_scope) = old_scope;
	}

	return ret;
}

/* {{{ */
static zval *eos_datastructure_struct_object_write_property(zend_object *object, zend_string *member, zval *value, void **cache_slot)
{
        zval *retval = NULL;

	if(eos_datastructures_struct_property_exists(object, member)) {
		/* This exists */
		retval = (zend_get_std_object_handlers())->write_property(object, member, value, cache_slot);
	} else {
		zend_throw_exception_ex(zend_ce_type_error, 0,
			"Name %s provided is not a property in struct %s",
			ZSTR_VAL(member), object->ce->name->val);
	}

        return retval;
}
/* }}} */

/* {{{ */
static zval *eos_datastructure_struct_object_read_property(zend_object *object, zend_string *member, int type, void **cache_slot, zval *rv)
{
	zval *retval = NULL;

	/* mangle scope so we can get to private/protected */
	zend_class_entry *old_scope = EG(fake_scope);
	EG(fake_scope) = object->ce;

	retval= (zend_get_std_object_handlers())->read_property(object, member, type, cache_slot, rv);

	/* return our scope to normal */
	EG(fake_scope) = old_scope;

	return retval;
}
/* }}} */

/* {{{ */
static void eos_datastructure_struct_object_unset_property(zend_object *object, zend_string *member, void **cache_slot)
{
	if(eos_datastructures_struct_property_exists(object, member)) {
		zend_throw_exception_ex(zend_ce_type_error, 0,
			"Name %s provided is a property in struct %s and cannot be unset",
			ZSTR_VAL(member), object->ce->name->val);
	} else {
		/* we shouldn't ever hit this, but just in case */
		(zend_get_std_object_handlers())->unset_property(object, member, cache_slot);
	}
}
/* }}} */


/* {{{ */
static zend_function *eos_datastructure_struct_object_get_method(zend_object **object, zend_string *method, const zval *key)
{
        zend_error_noreturn(E_ERROR, "Object does not support method calls");
}
/* }}} */


/* ----------------------------------------------------------------
    Eos\DataStructures\Struct Definition and registration
------------------------------------------------------------------*/

/* {{{ class methods */
static const zend_function_entry eos_datastructures_struct_methods[] = {
	PHP_ME(EosDataStructuresStruct, __construct, Struct___construct_args, ZEND_ACC_PUBLIC | ZEND_ACC_CTOR)
	ZEND_FE_END
};
/* }}} */

/* {{{ PHP_MINIT_FUNCTION */
PHP_MINIT_FUNCTION(eos_datastructures_struct)
{
	zend_class_entry ce;

	memcpy(&eos_datastructures_struct_object_handlers,
		   zend_get_std_object_handlers(),
		   sizeof(zend_object_handlers));

	eos_datastructures_struct_object_handlers.write_property = eos_datastructure_struct_object_write_property;
	eos_datastructures_struct_object_handlers.read_property = eos_datastructure_struct_object_read_property;
	eos_datastructures_struct_object_handlers.get_property_ptr_ptr = NULL;
	eos_datastructures_struct_object_handlers.get_method = eos_datastructure_struct_object_get_method;
	eos_datastructures_struct_object_handlers.unset_property = eos_datastructure_struct_object_unset_property;

	INIT_NS_CLASS_ENTRY(ce,  EOS_DATASTRUCTURES_NAMESPACE, "Struct", eos_datastructures_struct_methods);
	ce.create_object = eos_datastructures_struct_create_object;
	ce_eos_datastructures_struct = zend_register_internal_class(&ce);

	return SUCCESS;
}
/* }}} */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


src/eos_datastructures.c
/*
  +----------------------------------------------------------------------+
  | For PHP Version 7+                                                   |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015 Elizabeth M Smith                                 |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Author: Elizabeth M Smith <auroraeosrose@gmail.com>                  |
  +----------------------------------------------------------------------+
*/

#ifdef HAVE_CONFIG_H
#	include "config.h"
#endif

#include <php.h>
#include <ext/standard/info.h>

#include "php_eos_datastructures.h"
#include "php_eos_datastructures_internal.h"

/* {{{ PHP_MINIT_FUNCTION
 */
PHP_MINIT_FUNCTION(eos_datastructures)
{
	PHP_MINIT(eos_datastructures_enum)(INIT_FUNC_ARGS_PASSTHRU);
	PHP_MINIT(eos_datastructures_struct)(INIT_FUNC_ARGS_PASSTHRU);
	return SUCCESS;
}
/* }}} */

/* {{{ PHP_MINFO_FUNCTION */
PHP_MINFO_FUNCTION(eos_datastructures)
{
	php_info_print_table_start();
	php_info_print_table_colspan_header(2, "Eos Datastructures");
	php_info_print_table_row(2, "Extension", "enabled");
	php_info_print_table_row(2, "Type",
#ifdef COMPILE_DL_EOS_DATASTRUCTURES
		"compiled as dynamic module"
#else
		"compiled as static module"
#endif
        );
        php_info_print_table_row(2, "Extension Version", PHP_EOS_DATASTRUCTURES_VERSION);
	php_info_print_table_colspan_header(2, "DataTypes Available");
	php_info_print_table_row(2, "Enum Support", "enabled");
	php_info_print_table_row(2, "Struct Support", "enabled");
	php_info_print_table_end();
}
/* }}} */

/* {{{ datastructures_module_entry */
zend_module_entry eos_datastructures_module_entry = {
	STANDARD_MODULE_HEADER,
	"eos_datastructures",
	NULL,
	PHP_MINIT(eos_datastructures),
	NULL,
	NULL,
	NULL,
	PHP_MINFO(eos_datastructures),
	PHP_EOS_DATASTRUCTURES_VERSION,
	STANDARD_MODULE_PROPERTIES
};
/* }}} */

#ifdef COMPILE_DL_EOS_DATASTRUCTURES
	ZEND_GET_MODULE(eos_datastructures)
#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */

php_eos_datastructures_api.h
/*
  +----------------------------------------------------------------------+
  | For PHP Version 8                                                    |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015-2017 Elizabeth M Smith                            |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Authors: Elizabeth M Smith <auroraeosrose@gmail.com>                 |
  |          Swen Zanon <swen.zanon@geoglis.de>                          |
  +----------------------------------------------------------------------+
*/

#ifndef PHP_EOS_DATASTRUCTURES_API_H
#define PHP_EOS_DATASTRUCTURES_API_H

/* Public C api for use by other extensions */

#include <php.h>

#define PHP_EOS_DATASTRUTCURES_API_VERSION "1.0.1"

zend_class_entry *php_eos_datastructures_get_enum_ce(void);
long php_eos_datastructures_get_enum_value(zval* enumclass);
void php_eos_datastructures_set_enum_value(zval* enumclass, long value);
zend_bool php_eos_datastructures_check_value(zend_class_entry *ce, long value);

/* C apis for enums will go here */

#endif /* PHP_EOS_DATASTRUCTURES_API_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


src/php_eos_datastructures.h
/*
  +----------------------------------------------------------------------+
  | For PHP Version 8                                                    |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015-2017 Elizabeth M Smith                            |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Author: Elizabeth M Smith <auroraeosrose@gmail.com>                  |
  +----------------------------------------------------------------------+
*/

#ifndef PHP_EOS_DATASTRUCTURES_EXT_H
#define PHP_EOS_DATASTRUCTURES_EXT_H

#include <php.h>

extern zend_module_entry eos_datastructures_module_entry;
#define phpext_eos_datastructures_ptr &eos_datastructures_module_entry

#define PHP_EOS_DATASTRUCTURES_VERSION "1.0.0-dev"
#define EOS_DATASTRUCTURES_NAMESPACE "Eos\\DataStructures"

#ifdef PHP_WIN32
#   define PHP_EOS_DATASTRUCTURES_API __declspec(dllexport)
#elif defined(__GNUCC__) && __GNUC__ >= 4
#   define PHP_EOS_DATASTRUCTURES_API extern __attribute__ ((visibility("default")))
#else
#   define PHP_EOS_DATASTRUCTURES_API extern
#endif

#endif /* PHP_EOS_DATASTRUCTURES_EXT_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


src/php_eos_datastructures_internal.h
/*
  +----------------------------------------------------------------------+
  | For PHP Version 8                                                    |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015-2017 Elizabeth M Smith                            |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Author: Elizabeth M Smith <auroraeosrose@gmail.com>                  |
  +----------------------------------------------------------------------+
*/

#ifndef PHP_EOS_DATASTRUCTURES_INTERNAL_H
#define PHP_EOS_DATASTRUCTURES_INTERNAL_H

#include <php.h>

/* Classes to register */
PHP_MINIT_FUNCTION(eos_datastructures_enum);
PHP_MINIT_FUNCTION(eos_datastructures_struct);

#endif /* PHP_EOS_DATASTRUCTURES_INTERNAL_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


src/enum.c
/*
  +----------------------------------------------------------------------+
  | For PHP Version 8                                                    |
  +----------------------------------------------------------------------+
  | Copyright (c) 2015-2017 Elizabeth M Smith                            |
  +----------------------------------------------------------------------+
  | http://www.opensource.org/licenses/mit-license.php  MIT License      |
  | Also available in LICENSE                                            |
  +----------------------------------------------------------------------+
  | Author: Elizabeth M Smith <auroraeosrose@gmail.com>                  |
  +----------------------------------------------------------------------+
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <php.h>
#include <zend_exceptions.h>

#include "php_eos_datastructures.h"

zend_class_entry *ce_eos_datastructures_enum;
static zend_object_handlers eos_datastructures_enum_object_handlers;

/* enum object */
typedef struct _eos_datastructures_enum_object {
	zend_long   value;
	HashTable  *elements;
	zend_object std;
} eos_datastructures_enum_object;

#define EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(zv) ((eos_datastructures_enum_object*) \
	(((char*)Z_OBJ_P(zv)) - XtOffsetOf(eos_datastructures_enum_object, std)))

static int eos_datastructures_enum_apply_set(zend_long *option, int num_args, va_list args, zend_hash_key *hash_key);

/* ----------------------------------------------------------------
    Eos\Datastructures\Enum C API
------------------------------------------------------------------*/

/* {{{ function to take a zval** enum instance and give you back the long value */
PHP_EOS_DATASTRUCTURES_API zend_long php_eos_datastructures_get_enum_value(zval* enumclass)
{
	eos_datastructures_enum_object *enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(enumclass);
	return enum_object->value;
}
/* }}} */

/* {{{ function to take long and stick it in an enum class and validates it */
PHP_EOS_DATASTRUCTURES_API void php_eos_datastructures_set_enum_value(zval* enumclass, zend_long value)
{
	zend_bool found = 0;
	eos_datastructures_enum_object *enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(enumclass);

	/* handle the "easy" case of a long */
	zend_hash_apply_with_arguments(enum_object->elements,
		(apply_func_args_t)eos_datastructures_enum_apply_set, 2, &value, &found);

	if(found) {
		enum_object->value = value;
		return;
	} else {
		zend_throw_exception_ex(zend_ce_type_error, 0,
			"Value %d provided is not a const in enum %s",
			value, enum_object->std.ce->name->val);
	}
}
/* }}} */

/* {{{ function to validate a long against a ce */
PHP_EOS_DATASTRUCTURES_API zend_bool php_eos_datastructures_check_value(zend_class_entry *ce, zend_long value)
{
	zval *val;
	zend_bool return_value = 0;

	ZEND_HASH_FOREACH_VAL(&ce->constants_table, val) {
            if (Z_TYPE_P(val) == IS_PTR) {
                val = (zval *) Z_PTR_P(val);
            }
            if(Z_LVAL_P(val) == value) {
                    return_value = 1;
                    break;
            }
	} ZEND_HASH_FOREACH_END();

	if(!return_value) {
		zend_throw_exception_ex(zend_ce_type_error, 0,
			"Value %d provided is not a const in enum %s",
			value, ce->name->val);
	}
	return return_value;
}
/* }}} */

/* {{{ allows access to the enum class entry via other extensions */
PHP_EOS_DATASTRUCTURES_API zend_class_entry *php_eos_datastructures_get_enum_ce(void)
{
	return ce_eos_datastructures_enum;
}
/* }}} */

/* ----------------------------------------------------------------
    Eos\DataStructures\Enum class API
------------------------------------------------------------------*/

ZEND_BEGIN_ARG_INFO(Enum___construct_args, ZEND_SEND_BY_VAL)
	ZEND_ARG_INFO(0, value)
ZEND_END_ARG_INFO()

/* {{{ proto void Enum->__construct(mixed value)
                  constructor for enum class, string can be a special case
                  anything else is cast and we attempt to assign */
PHP_METHOD(EosDataStructuresEnum, __construct)
{
	eos_datastructures_enum_object *enum_object;
	zend_bool found = 0;
	zval *name = NULL;
	zend_long constant_value;

	/* Pass 1 - check for a string value WITHOUT juggling */
        ZEND_PARSE_PARAMETERS_START(1,1)
                Z_PARAM_ZVAL(name)
        ZEND_PARSE_PARAMETERS_END();

	enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(getThis());

	/* If we are given a string, look up a constant name */
	if(Z_TYPE_P(name) == IS_STRING) {
		zval *found_zval = zend_hash_find(enum_object->elements, Z_STR_P(name));
		if(found_zval) {
			enum_object->value = Z_LVAL_P(found_zval);
			return;
		}
		/* If this is NOT a numeric string, bail, otherwise continue */
		if(!is_numeric_string(Z_STRVAL_P(name), Z_STRLEN_P(name), NULL, NULL, 0)) {
			zend_throw_exception_ex(zend_ce_type_error, 0,
				"Name %s provided is not a const in enum %s",
				Z_STRVAL_P(name), enum_object->std.ce->name->val);
			return;
		}
	}

	/* Pass 2, so we didn't get a string, parse it hard as a long */
	if (FAILURE == zend_parse_parameters_throw(ZEND_NUM_ARGS(), "l", &constant_value)) {
		return;
	}
	/* handle the "easy" case of a long */
	zend_hash_apply_with_arguments(enum_object->elements,
		(apply_func_args_t)eos_datastructures_enum_apply_set, 2, &constant_value, &found);

	if(found) {
		enum_object->value = constant_value;
		return;
	}

	zend_throw_exception_ex(zend_ce_type_error, 0,
		"Value %d provided is not a const in enum %s",
		constant_value, enum_object->std.ce->name->val);
}
/* }}} */

ZEND_BEGIN_ARG_INFO(Enum_getName_args, ZEND_SEND_BY_VAL)
ZEND_END_ARG_INFO()

/* {{{ proto string Enum->getName()
get member name of current enum value */
PHP_METHOD(EosDataStructuresEnum, getName)
{
	eos_datastructures_enum_object *enum_object;
	zend_ulong num_idx;
	zend_string *str_idx;
	zval *entry;
	//zval tmp;

	ZEND_PARSE_PARAMETERS_NONE();

	enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(getThis());

	ZEND_HASH_FOREACH_KEY_VAL(enum_object->elements, num_idx, str_idx, entry) {
		if(Z_LVAL_P(entry) == enum_object->value) {
			RETURN_STR_COPY(str_idx);
		}
	} ZEND_HASH_FOREACH_END();
}
/* }}} */

ZEND_BEGIN_ARG_INFO(Enum_getElements_args, ZEND_SEND_BY_VAL)
ZEND_END_ARG_INFO()

/* {{{ proto array Enum->getElements()
                   get array of available name => value pairs */
PHP_METHOD(EosDataStructuresEnum, getElements)
{
	eos_datastructures_enum_object *enum_object;

	ZEND_PARSE_PARAMETERS_NONE();

	enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(getThis());

	array_init(return_value);
	zend_hash_copy(Z_ARRVAL_P(return_value), enum_object->elements, zval_add_ref);
}
/* }}} */

/* ----------------------------------------------------------------
    Eos\DataStructures\Enum Object management
------------------------------------------------------------------*/

/* {{{ */
static void eos_datastructures_enum_free_obj(zend_object *object)
{
	eos_datastructures_enum_object *intern = (eos_datastructures_enum_object*)
		((char*)object - XtOffsetOf(eos_datastructures_enum_object, std));

	if(!intern) {
            return;
	}
	intern->value = 0;
	zend_hash_destroy(intern->elements);
	FREE_HASHTABLE(intern->elements);

	zend_object_std_dtor(&intern->std);
}
/* }}} */

/* {{{ */
static zend_object* eos_datastructures_enum_obj_ctor(zend_class_entry *ce, eos_datastructures_enum_object **intern)
{
	eos_datastructures_enum_object *object = ecalloc(1, sizeof(eos_datastructures_enum_object)
                                                        + zend_object_properties_size(ce));

	object->value = 0;
	zend_update_class_constants(ce);

	zend_object_std_init(&object->std, ce);
	object->std.handlers = &eos_datastructures_enum_object_handlers;
	*intern = object;

	return &object->std;
}
/* }}} */

/* {{{ */
static zend_object* eos_datastructures_enum_create_object(zend_class_entry *ce)
{
	eos_datastructures_enum_object *intern = NULL;

	zend_object *return_value = eos_datastructures_enum_obj_ctor(ce, &intern);
	ALLOC_HASHTABLE(intern->elements);
	zend_hash_init(intern->elements, 8, NULL, NULL, 0);

	int count = zend_hash_num_elements(&ce->constants_table);
	if (count > 0) {
		zend_string *key;
		zval *value;

		ZEND_HASH_FOREACH_STR_KEY_PTR(&ce->constants_table, key, value) {
			if(Z_TYPE_P(value) != IS_LONG) {
				php_error(E_WARNING, "Constant %s is being cast to an integer in Enum subclass %s", key->val, ce->name->val);
				convert_to_long(value);
			}

			zend_hash_add(intern->elements, key, value);
		} ZEND_HASH_FOREACH_END();
	}

	object_properties_init(&(intern->std), ce);
	return return_value;
}
/* }}} */

/* {{{ */
static zend_object* eos_datastructures_enum_clone_obj(zend_object *object) 
{
        eos_datastructures_enum_object *new_enum;
        // copied from eos_datastructures_enum_free_obj():
        eos_datastructures_enum_object *old_enum = (eos_datastructures_enum_object*)
                ((char*)object - XtOffsetOf(eos_datastructures_enum_object, std));
        zend_object *return_value = eos_datastructures_enum_obj_ctor(object->ce, &new_enum);

        new_enum->value = old_enum->value;
        new_enum->elements = zend_array_dup(old_enum->elements);

        zend_objects_clone_members(&new_enum->std, &old_enum->std);

        return return_value;
}
/* }}} */

/* {{{ helper function for finding a valid enum value via long */
static int eos_datastructures_enum_apply_set(zend_long *option, int num_args, va_list args, zend_hash_key *hash_key) 
{
	zend_long *value = va_arg(args, zend_long*);
	zend_bool *found = va_arg(args, zend_bool*);

	if(*value == *option) {
		*found = 1;
		return ZEND_HASH_APPLY_STOP;
	}

	return ZEND_HASH_APPLY_KEEP;
}
/* }}} */

/* {{{ */
/*
 * HINTS:
 * The get() and set() object handlers have been removed.
 * see https://raw.githubusercontent.com/php/php-src/PHP-8.0/UPGRADING.INTERNALS
 */
//static void eos_datastructures_enum_object_set(zval *zobject, zval *value)
//{
//	eos_datastructures_enum_object *enum_object = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(zobject);
//	zend_bool found = 0;
//	zend_long lvalue;
//	zval *juggled;
//
//	if (Z_TYPE_P(value) == IS_OBJECT && Z_OBJ_HANDLER_P(value, get)) {
//		zval rv;
//		juggled = Z_OBJ_HANDLER_P(value, get)(value, &rv);
//	} else {
//		juggled = value;
//	}
//
//	/* if we have a string, do a zend_hash_find FIRST */
//	if (Z_TYPE_P(juggled) == IS_STRING) {
//		zval *found_zval =zend_hash_find(enum_object->elements, Z_STR_P(juggled));
//		if(found_zval) {
//			enum_object->value = Z_LVAL_P(found_zval);
//			return;
//		}
//		/* If this is NOT a numeric string, bail, otherwise continue */
//		if(!is_numeric_string(Z_STRVAL_P(juggled), Z_STRLEN_P(juggled), NULL, NULL, 0)) {
//			zend_throw_exception_ex(zend_ce_type_error, 0,
//				"Name %s provided is not a const in enum %s",
//				Z_STRVAL_P(juggled), enum_object->std.ce->name->val);
//			return;
//		}
//	}
//
//	/* If juggled is long or a string, leave them alone, otherwise juggle */
//	if (Z_TYPE_P(juggled) != IS_LONG) {
//		convert_to_long(juggled);
//	}
//	lvalue = Z_LVAL_P(juggled);
//
//	/* handle the "easy" case of a long */
//	zend_hash_apply_with_arguments(enum_object->elements,
//		(apply_func_args_t)eos_datastructures_enum_apply_set, 2, &lvalue, &found);
//
//	if(found) {
//		enum_object->value = lvalue;
//		return;
//	} else {
//		zend_throw_exception_ex(zend_ce_type_error, 0,
//		"Value %d provided is not a const in enum %s",
//		lvalue, enum_object->std.ce->name->val);
//	}
//}
/* }}} */

/* {{{ */
static int eos_datastructures_enum_cast_object(zend_object *readobj, zval *writeobj, int type)
{
        eos_datastructures_enum_object *enum_object = (eos_datastructures_enum_object*)
                ((char*)readobj - XtOffsetOf(eos_datastructures_enum_object, std));
        
	ZVAL_LONG(writeobj, enum_object->value);
	convert_to_explicit_type(writeobj, type);
	return SUCCESS;
}
/* }}} */

/* {{{ */
static int eos_datastructures_enum_compare_objects(zval *z1, zval *z2)
{
	if (Z_TYPE_P(z1) == IS_OBJECT && Z_TYPE_P(z2) == IS_OBJECT &&
		instanceof_function(Z_OBJCE_P(z1), ce_eos_datastructures_enum) &&
		instanceof_function(Z_OBJCE_P(z2), ce_eos_datastructures_enum)) {

			eos_datastructures_enum_object *enum1 = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(z1);
			eos_datastructures_enum_object *enum2 = EOS_DATASTRUCTURES_ENUM_FETCH_OBJ(z2);
                        
			return (enum1->value == enum2->value) ? 0 : ((enum1->value < enum2->value) ? -1 : 1);
		}
        
        if (Z_TYPE_P(z1) != IS_OBJECT ||
            Z_TYPE_P(z2) != IS_OBJECT ||
            Z_OBJ_HT_P(z1)->compare != Z_OBJ_HT_P(z2)->compare) {
		return zend_std_compare_objects(z1, z2);
	}

	return 1;
}
/* }}} */

/* {{{ */
static HashTable* eos_datastructures_enum_object_get_debug_info(zend_object *object, int *is_temp)
{
	HashTable *debug_info, *std_props;
	zval elements, value;
        eos_datastructures_enum_object *enum_object = (eos_datastructures_enum_object*)
                ((char*)object - XtOffsetOf(eos_datastructures_enum_object, std));
        
	std_props = zend_std_get_properties(object);
	debug_info = zend_array_dup(std_props);

	array_init(&elements);
	zend_hash_copy(Z_ARRVAL(elements), enum_object->elements, zval_add_ref);
	zend_hash_str_add(debug_info, "__elements", sizeof("__elements") -1, &elements);

	ZVAL_LONG(&value, enum_object->value);
	zend_hash_str_add(debug_info, "__value", sizeof("__value")-1, &value);

	*is_temp = 1;
	return debug_info;
}
/* }}} */

/* ----------------------------------------------------------------
    Eos\DataStructures\Enum Definition and registration
------------------------------------------------------------------*/

/* {{{ class methods */
static const zend_function_entry eos_datastructures_enum_methods[] = {
	PHP_ME(EosDataStructuresEnum, __construct, Enum___construct_args, ZEND_ACC_PUBLIC | ZEND_ACC_CTOR)
	PHP_ME(EosDataStructuresEnum, getName, Enum_getName_args, ZEND_ACC_PUBLIC)
	PHP_ME(EosDataStructuresEnum, getElements, Enum_getElements_args, ZEND_ACC_PUBLIC)
	ZEND_FE_END
};
/* }}} */

/* {{{ PHP_MINIT_FUNCTION */
PHP_MINIT_FUNCTION(eos_datastructures_enum)
{
	zend_class_entry ce;

	memcpy(&eos_datastructures_enum_object_handlers,
		   zend_get_std_object_handlers(),
		   sizeof(zend_object_handlers));

	eos_datastructures_enum_object_handlers.offset = XtOffsetOf(eos_datastructures_enum_object, std);
	eos_datastructures_enum_object_handlers.free_obj = eos_datastructures_enum_free_obj;
	eos_datastructures_enum_object_handlers.clone_obj = eos_datastructures_enum_clone_obj;
	eos_datastructures_enum_object_handlers.cast_object = eos_datastructures_enum_cast_object;
	eos_datastructures_enum_object_handlers.compare = eos_datastructures_enum_compare_objects;
	eos_datastructures_enum_object_handlers.get_debug_info = eos_datastructures_enum_object_get_debug_info;
        /*
        * The get() and set() object handlers have been removed.
        * see https://raw.githubusercontent.com/php/php-src/PHP-8.0/UPGRADING.INTERNALS
        */
//	eos_datastructures_enum_object_handlers.set = eos_datastructures_enum_object_set;

	INIT_NS_CLASS_ENTRY(ce,  EOS_DATASTRUCTURES_NAMESPACE, "Enum", eos_datastructures_enum_methods);
	ce.create_object = eos_datastructures_enum_create_object;
	ce_eos_datastructures_enum = zend_register_internal_class(&ce);
	ce_eos_datastructures_enum->ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;

	return SUCCESS;
}
/* }}} */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
